OpenScienceJournal/whistlepunk_library/src/main/java/com/google/android/apps/forscience/ble/MkrSciBleManager.java

```diff
@@ -0,0 +1,426 @@
+            boolean unsubscribe = false;
+            synchronized (mListeners) {
+                List<Listener> listeners = mListeners.get(characteristicUuid);
+                if (listeners != null) {
+                    listeners.remove(listener);
+                    if (listeners.size() == 0) {
+                        mListeners.remove(characteristicUuid);
+                        unsubscribe = true;
+                    }
+                }
+            }
+            if (unsubscribe) {
+                enqueueGattAction(() -> {
+                    BluetoothGattCharacteristic c = getCharacteristic(characteristicUuid);
+                    if (c != null) {
+                        mGatt.setCharacteristicNotification(c, true);
+                        BluetoothGattDescriptor d = c.getDescriptor(NOTIFICATION_DESCRIPTOR);
+                        d.setValue(BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE);
+                        mGatt.writeDescriptor(d);
+                    }
+                });
+            }
+        }
+
+        private boolean hasSubscribers() {
+            synchronized (mListeners) {
+                return mListeners.size() > 0;
+            }
+        }
+
+        private BluetoothGattCharacteristic getCharacteristic(String uuid) {
+            for (BluetoothGattCharacteristic aux : mCharacteristics) {
+                if (Objects.equals(uuid, aux.getUuid().toString())) {
+                    return aux;
+                }
+            }
+            return null;
+        }
+
+        private void enqueueGattAction(Runnable action) {
+            synchronized (mGattActions) {
+                if (mReadyForAction && !mBusy) {
+                    mBusy = true;
+                    action.run();
+                } else {
+                    mGattActions.add(action);
+                }
+            }
+        }
+
+        private void onGattActionCompleted() {
+            synchronized (mGattActions) {
+                if (mReadyForAction && mGattActions.size() > 0) {
+                    mBusy = true;
+                    mGattActions.remove(0).run();
+                } else {
+                    mBusy = false;
+                }
+            }
+        }
+
+        @Override
+        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
+            if (status == BluetoothGatt.GATT_SUCCESS
+                    && newState == BluetoothProfile.STATE_CONNECTED) {
+                mGatt = gatt;
+                mCharacteristics.clear();
+                mGatt.discoverServices();
+            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
+                mReadyForAction = false;
+                gatt.disconnect();
+            }
+        }
+
+        @Override
+        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
+            BluetoothGattService service = gatt.getService(UUID.fromString(SERVICE_UUID));
```

Comment:
At this point, is gatt always the same as mGatt? I notice it uses gatt on line 225 and mGatt on line 231 and was wondering if that is necessary. If so, can you add a comment explaining.

Reply:
Yes, they are the same at that point. I uniformed them by using mGatt in both cases. Thank you.

Reply:
Thanks. Can you explain in more detail exactly what this commit will fix? We've seen problems where some devices (for example, a Chromebook) would never get data from any of the arduino science kit sensors. We've seen other problems where some devices (for examples, a Chrome OS tablet) will get some data and then after a few minutes the data will stop. Will this commit fix either of these two issues?

Reply:
Latest commits (since a6b0a6d on) solve an incompatibility with some Android 5.x and 6.x devices that have been reported they can't connect the Arduino board before this fix.

Current commits do not fix the data stop problem reported in other situations. That problem is more related to the board than to the app: when the board battery is low, there's no enough energy to power its circuits. I am working to another solution for this, implementing a low battery notification in the app, but this will require some time and a new firmware version on the board as well. Compatibility issues with Android 5/6 are far more urgent.